src/lib.rs

```diff
@@ -165,50 +190,72 @@ impl GpuProfiler {
     ///
     /// Scopes can be arbitrarily nested.
     ///
-    /// May create new wgpu query objects (which is why it needs a [`wgpu::Device`] reference)
+    /// May create new wgpu query objects (which is why it needs a [`wgpu::Device`] reference).
+    ///
+    /// If encoder/pass timer queries are disabled respectively, a scope will still be opened,
+    /// but the resulting [`GpuTimerScopeResult`] will have `None` for its `time` field.
     ///
     /// See also [`wgpu_profiler!`], [`GpuProfiler::end_scope`]
     #[track_caller]
     pub fn begin_scope<Recorder: ProfilerCommandRecorder>(&mut self, label: &str, encoder_or_pass: &mut Recorder, device: &wgpu::Device) {
-        if (encoder_or_pass.is_pass() && self.enable_pass_timer) || (!encoder_or_pass.is_pass() && self.enable_encoder_timer) {
+        // Open a scope even if we don't perform a query, this makes error handling a bit more consistent and
+        // preserves scope hierarchy even if certain kinds of scopes are disabled.
+        let start_query = if (encoder_or_pass.is_pass() && self.enable_pass_timer) || (!encoder_or_pass.is_pass() && self.enable_encoder_timer) {
```

Comment:
Does this mean that we still open scopes even if GpuProfiler is disabled overall? I would personally prefer to avoid any excess work in this case.

Reply:
I'm a bit torn on this. It's obviously nice for testing, but I figured it's also neat to get the pass scopes even if profiling is only supported on encoders (which happens on apple silicon). In those cases we'd now get `GpuProfilingScope::time` set to `None`

Right now, the profiler actually can't be disabled, only `enable_debug_marker` is exposed. I think we should provide a runtime way to make begin_scope/end_scope be a no-op. If that's used then we'd also not expect `GpuProfilingScope` 🤔
Not sure yet either what's the interface for that: If we disable the profiler proper, then we should also dispose all the buffers etc. or not create them in the first place if that's done on creation (how?). But in that case we'd not distinguish pass & encoder scopes.
What do you think?

Reply:
Ohhhh, I'm blocked on updating wgpu so I'm using a version that still has a public `enable_timer` field. So I just kind of assumed there was still a way for users to toggle this. Would love to see some equivalent feature in newer versions but it seems like it isn't a concern for this PR.

> If that's used then we'd also not expect `GpuProfilingScope`

I'm a bit lost on this point.

> Not sure yet either what's the interface for that: If we disable the profiler proper, then we should also dispose all the buffers etc. or not create them in the first place if that's done on creation (how?).

I guess `GpuProfiler::new` could take an enabled `bool` and then have a method for setting that value where it can do cleanup if it is disabled. I'm not sure if there are any other details to handle. If necessary, any internal state that relies on the profiler having buffers/being enabled be wrapped in an `enabled: Option<State>`, but I'm not sure that this is necessary in the current setup. 

> But in that case we'd not distinguish pass & encoder scopes.

:+1:

Reply:
looking at this again I found this too convoluted after all: I went back to making GpuProfiler scopes properly disabled as before.
Will expose disabling the profiler in a controlled fashion in a followup!